### DISCUSSION TITLE
Display custom html within AR

### DISCUSSION THREAD

**Discussion by Pratik-499** [google/model-viewer#2708] [Q&A]
### Description

I want to display an "Add to cart" button within AR. I'm using "glb" file format for 3d file for both android and ios. 
It works in Android but not working in ios. 

How can we display a custom button within AR that support both ios and android?

`<model-viewer src="https://cdn.shopify.com/s/files/1/0456/5509/7511/files/Chair.glb?v=1628335986" ios-src="https://cdn.shopify.com/s/files/1/0456/5509/7511/files/Wooden_barrel.usdz?v=1629738829" alt="A 3D model"  ar ar-modes="webxr scene-viewer quick-look" auto-rotate camera-controls shadow-intensity="1" ar-placement="wall">   
 <button type="button" class="btn-ar-addToCart">
    <span class="btn-txt">Add to cart</span>
 </button>        
 </model-viewer>
`

![Image of Yaktocat](https://i.stack.imgur.com/peuvg.png)

### Version

- model-viewer: vX.X.X

### OS

- [ ] Android
- [x] iOS

### AR

- [ ] WebXR
- [ ] SceneViewer
- [x] QuickLook

**Comment by elalish** [comment]
Unfortunately until Apple supports WebXR, there isn't a consistent way to do this. However, you can add query params to your `ios-src` to get most of what you want: https://developer.apple.com/documentation/arkit/adding_an_apple_pay_button_or_a_custom_action_in_ar_quick_look

**Reply by rawwerks** [reply]
@Pratik-499 - did you ever get this working? 

if so, would you be willing to share your code?

if not, maybe this is related: https://github.com/google/model-viewer/discussions/3425

**Reply by milesgreen** [reply]
To display a banner in AR on iOS you need to append specific QuickLook parameters to the USDZ path as # URL fragment in ModelViewer's `ios-src` attribute.

EG:
`ios-src="https://example.com/model.usdz#applePayButtonType=pay"`

Apple has more documentation about QuickLook parameters:
https://developer.apple.com/documentation/arkit/adding_an_apple_pay_button_or_a_custom_action_in_ar_quick_look

Per the Apple documentation, valid parameters for configuring the appearance of the QuickLook banner include:

**`applePayButtonType`**
Predefined Apple Pay button to display in banner. Choice of: `plain` / `pay` / `buy` / `check-out` / `book` / `donate` / `subscribe`

**`callToAction`**
Custom button label to display in banner, rather than an Apple Pay button

**`checkoutTitle`**
String to display as the banner title

**`checkoutSubtitle`**
String. Line of copy to display under the Title in the banner.

**`price`**
String. Price to be displayed in banner.

EG:
`ios-src="https://example.com/model.usdz#callToAction=Add%20to%20cart&checkoutTitle=My%20Title&checkoutSubtitle=Description%20Text&price=%2499"`

Note that values should be correctly URL escaped if needed.

To respond to clicks on the QuickLook banner, listen for the `quick-look-button-tapped` ModelViewer event and respond accordingly:
https://modelviewer.dev/docs/index.html#entrydocs-augmentedreality-events-quickLookButtonTapped

**Reply by rawwerks** [reply]
thanks for the quick save @milesgreen ! (as usual). 

clarifying question - i assume this approach would preclude the use of the auto-generated usdz. is this correct? or is there a way to point ios-src to the html of the auto-generated usdz?

**Reply by milesgreen** [reply]
No problem! 

If using auto-generated USDZ, instead simply add these # parameters to the end of the GLB path of the `src` parameter.

**Reply by milesgreen** [reply]
Note that passing # parameters through auto-generated USDZs only works for Safari iOS 15.5 and up. Prior to that, Safari didn't handle it well and either ignored the parameters or, worse, could actually block the 3D model from ever loading in QuickLook. So it's worth checking for that, and having a graceful fallback, or directing users to upgrade.

You'll probably want to use at least ModelViewer v2.1.0 as that also includes a number of related iOS fixes:

https://github.com/google/model-viewer/pull/3434
https://github.com/google/model-viewer/pull/3784
https://github.com/google/model-viewer/pull/3850

**Reply by rawwerks** [reply]
is this is the right way to listen to the event?
```
<model-viewer id="mv" otherstuff></model-viewer>
<script>
    document.querySelector("#mv").addEventListener('quick-look-button-tapped', (event) => {
      window.location.assign('https://my.link.to.cart');
    });
</script>
```
update: this seems to work for me, just want to have someone check my work since there isn't a specific example on the docs for `quick-look-button-tapped`

**Reply by elalish** [reply]
Yep, just like any other event.

**Comment by rawwerks** [comment]
on this topic of cross-compatible in-AR buy buttons - anyone have the CSS for the default apple quicklook buy button, so that i can have the same style on Android? (or any examples showing code that provides an in-AR buy button regardless of which ar mode is being used?)
